### PR TITLE
working with user requests

### PR DESCRIPTION
**Pull request by stadnytskyivp**

### CODE REVIEW

File: src/main/java/client/UserClient.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import io.restassured.response.Response;
+import io.restassured.specification.RequestSpecification;
+import org.apache.http.HttpStatus;
+
+import java.io.IOException;
+import java.util.List;
+
+public class UserClient extends BaseClient {
+    final private static String USER_ENDPOINT = "/v2/user/";
+    final private static String USER_LIST_ENDPOINT = USER_ENDPOINT + "createWithList";
```
Comment (andyyarych) [inline src/main/java/client/UserClient.java:15]: Remove the last "/" from USER_ENDPOINT and put it at the beginning of every new endpoint: `/createWithList`, `/createWithArray` etc

File: src/main/java/client/UserClient.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    final private static String USER_ARRAY_ENDPOINT = USER_ENDPOINT + "createWithArray";
+
+    @Step("Adding user to the store data base")
+    public static ResponseInfo postUser(User userToPost) throws IOException {
+        LOGGER.debug("sending request");
```
Comment (andyyarych) [inline src/main/java/client/UserClient.java:20]: Maybe, consider updating debug messages to something more informative, like: `"Sending POST request to /user"` etc

File: src/main/java/data/DataSet.java [hunk 3]
```diff
@@ -50,4 +51,15 @@ public static Order addingOrder() {
             .setComplete(true);
     }
 
+    public static User addingUser() {
+        return new User()
+            .setId(311L)
```
Comment (andyyarych) [inline src/main/java/data/DataSet.java]: Is it necessary to set user id in request?

File: src/test/java/petStoreTests/user/AddUserTest.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+
+    @Description("Verify that we are adding user to the store data base")
+    @Test
+    public void addUserTest() throws IOException {
+        LOGGER.info("START TEST add user to the store data base");
+        ResponseInfo response = UserClient.postUser(DataSet.addingUser());
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), String.valueOf(DataSet.addingUser().getId()));
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/AddUserTest.java]: Also add GET request for the specified user by id and verify that returned user is equal to the one you posted

File: src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+    @Description("Verify that we are adding array of users to the store data base")
+    @Parameters({"Users for adding"})
+    @Test(dataProvider = "testData")
+    public void addUserTest(User[] users) throws IOException {
+        LOGGER.info("START TEST add users to the store data base");
+        ResponseInfo response = UserClient.postUserArray(users);
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), DataSet.messageUnknownResponse().setMessage("ok").getMessage());
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java:33]: How do you verify that these users where actually created? Consider sending GET requests by each of users ids and verifying that they were actually added.

File: src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        User user3 = DataSet.addingUser().setId(103L);
+        User[] treeUsers = new User[]{user1, user2, user3};
+        User[] oneUser = new User[]{user1};
+        User[] duplicateUsers = new User[]{user1, user1};
+
+        return new Object[][]{
+            {treeUsers},
+            {oneUser},
+            {duplicateUsers},
+            {Collections.emptyList()}
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java]: This is invalid parameter and test will fail due to data mismatch

File: src/test/java/petStoreTests/user/CreateListOfUsersTest.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    @Parameters({"Users for adding"})
+    @Test(dataProvider = "testData")
+    public void addUserTest(List<User> users) throws IOException {
+        LOGGER.info("START TEST add users to the store data base");
+        ResponseInfo response = UserClient.postUserList(users);
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), DataSet.messageUnknownResponse().setMessage("ok").getMessage());
+        LOGGER.info("END TEST");
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/CreateListOfUsersTest.java]: Same here as in previous case: add verification that users were actually created

File: src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), DataSet.messageUnknownResponse().setMessage("ok").getMessage());
+        LOGGER.info("END TEST");
+    }
+
+    @DataProvider
+    public Object[][] testData() {
+        User user1 = DataSet.addingUser().setId(101L);
+        User user2 = DataSet.addingUser().setId(102L);
+        User user3 = DataSet.addingUser().setId(103L);
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java]: Again, is it necessary to add user ids? Won't they rewrite already existing users?

File: src/test/java/petStoreTests/user/DeleteUserByUsernameTest.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+        UserClient.postUser(DataSet.addingUser());
+        LOGGER.info("BEFORE TEST USER ADDED");
+
+        LOGGER.info("START TEST delete user from the store data base");
+        ResponseInfo response = UserClient.deleteUserByUsername(DataSet.addingUser().getUsername());
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), DataSet.addingUser().getUsername());
+        LOGGER.info("END TEST");
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/DeleteUserByUsernameTest.java]: How do you verify that user was actually deleted?

File: src/test/java/petStoreTests/user/DeleteUserNegativeTest.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+
+        LOGGER.info("BEFORE TEST making sure that the base can't delete non existing user");
+        UserClient.postUser(DataSet.addingUser());
+        UserClient.deleteUserByUsername(DataSet.addingUser().getUsername());
+        LOGGER.info("BEFORE TEST USER ADDED AND DELETED");
+
+        LOGGER.info("START TEST delete non existing user from the store data base");
+        Response response = UserClient.deleteNonExistingUser(DataSet.addingUser().getUsername());
+
+        Assert.assertTrue(response.asString().trim().isEmpty());
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/DeleteUserNegativeTest.java]: I don't really understand this assertion

File: src/test/java/petStoreTests/user/FindUserByUsernameTest.java [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+    public void findUserTest(User user) throws IOException {
+
+        LOGGER.info("BEFORE TEST ADD USER");
+        UserClient.postUser(user);
+        LOGGER.info("BEFORE TEST USER ADDED");
+
+        LOGGER.info("START TEST find user in the store data base");
+        User response = UserClient.getUserByUsername(user.getUsername());
+
+        Assert.assertTrue(response.getId() >= 0);
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/FindUserByUsernameTest.java]: I don't think this is valid assertion, verifying for non-null value should be enough

File: src/test/java/petStoreTests/user/FindUserNegativeTest.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+
+public class FindUserNegativeTest extends AbstractTest {
+
+    @Description("Verify that we are can find user in the store data base")
+    @Parameters({"User for adding and finding"})
+    @Test(dataProvider = "testData")
+    public void findUserTest(User user) throws IOException {
+
+        LOGGER.info("BEFORE TEST ADD USER");
+        UserClient.deleteUserByUsername(user.getUsername());
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/FindUserNegativeTest.java]: Logger says you are adding a user, but I see a call of `deleteUserByUsername()` method

File: src/test/java/petStoreTests/user/UpdateUserTest.java [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+    @Description("Verify that we are editing user data in the store data base")
+    @Parameters({"User for editing"})
+    @Test(dataProvider = "testData")
+    public void updateUserTest(User expectedUser) throws IOException {
+        LOGGER.info("BEFORE TEST ADDING USER");
+        UserClient.postUser(expectedUser);
+        LOGGER.info("BEFORE TEST USER ADDED");
+
+        LOGGER.info("START TEST edit user info in the store data base");
+        ResponseInfo response = UserClient.updateUser(expectedUser);
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/UpdateUserTest.java]: What is the point of updating the user with the same exact user?

**Review by andyyarych — CHANGES_REQUESTED** [review]

**Comment by andyyarych** [comment]
Also remove empty lines after class and methods declaration

File: src/main/java/data/DataSet.java (hunk 3 above)
Comment (stadnytskyivp) [reply src/main/java/data/DataSet.java]: @andyyarych no, not really. In FindUserByUsernameTest class I have test where I'm posting ID = null, and it's goes well.

File: src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java (hunk 8 above)
Comment (stadnytskyivp) [reply src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java]: @andyyarych I have added those Ids cuz they are rewriting existing users

File: src/test/java/petStoreTests/user/DeleteUserNegativeTest.java (hunk 10 above)
Comment (stadnytskyivp) [reply src/test/java/petStoreTests/user/DeleteUserNegativeTest.java]: @andyyarych when there is no user to delete i'm getting status code 404 and empty response. So here i'm waiting for it)

File: AllSuitesTestng.xml [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+            <class name="petStoreTests.user.FindUserByUsernameTest"/>
+            <class name="petStoreTests.user.FindUserNegativeTest"/>
+            <class name="petStoreTests.user.CreateArrayOfUsersTest"/>
+            <class name="petStoreTests.user.CreateListOfUsersTest"/>
+            <class name="petStoreTests.user.DeleteUserByUsernameTest"/>
+            <class name="petStoreTests.user.DeleteUserNegativeTest"/>
+            <class name="petStoreTests.user.UpdateUserTest"/>
+        </classes>
+    </test>
+</suite>
```
Comment (andyyarych) [inline AllSuitesTestng.xml]: No new line at the end of file

File: src/main/java/client/StoreClient.java [hunk 15]
```diff
@@ -76,7 +75,7 @@ public static ResponseInfo deleteOrderById(long orderId) throws IOException {
             .as(ResponseInfo.class);
     }
 
-    @Step("Trying of getting nonexistent order by ID {0}")
+    @Step("Sending GET request to /store/order by ID {0}")
     public static ResponseInfo getNonexistentOrderById(long orderId) throws IOException {
```
Comment (andyyarych) [inline src/main/java/client/StoreClient.java]: Step description should be more explanatory

File: src/main/java/client/StoreClient.java [hunk 16]
```diff
@@ -96,7 +95,7 @@ public static ResponseInfo getNonexistentOrderById(long orderId) throws IOExcept
             .as(ResponseInfo.class);
     }
 
-    @Step("Deleting order by ID {0}")
+    @Step("Sending DELETE request to /store/order by ID {0}")
```
Comment (andyyarych) [inline src/main/java/client/StoreClient.java]: Same here - if you delete here by non-existing id, explain it in a description

File: src/main/java/data/DataSet.java [hunk 17]
```diff
@@ -50,4 +50,15 @@ public static Order addingOrder() {
             .setComplete(true);
     }
 
+    public static User addingUser() {
+        return new User()
+            .setId(Long.parseLong(""+ReusableMethods.getRandomNumber()))
```
Comment (andyyarych) [inline src/main/java/data/DataSet.java]: Haven't we talked about not setting id at all when creating an user?

File: src/test/java/petStoreTests/user/AddUserTest.java [hunk 18]
```diff
@@ -0,0 +1,35 @@
...
+        LOGGER.info("START TEST add user to the store data base");
+        User user = DataSet.addingUser();
+        ResponseInfo response = UserClient.postUser(user);
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), String.valueOf(user.getId()));
+
+        User newUser = UserClient.getUserByUsername(user.getUsername());
+        Assert.assertEquals(newUser.getId(), user.getId());
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/AddUserTest.java]: Why not assert that both `user` and `newUser` objects are equal, and not only their usernames?

File: src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java [hunk 19]
```diff
@@ -0,0 +1,60 @@
...
+        ResponseInfo response = UserClient.postUserArray(users);
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), DataSet.messageUnknownResponse().setMessage("ok").getMessage());
+
+        if (users.length != 0){
+            for (User user : users) {
+                User actualUser = UserClient.getUserByUsername(user.getUsername());
+                Assert.assertEquals(actualUser.getUsername(), user.getUsername());
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/CreateArrayOfUsersTest.java]: Is this the only assertion that can be applied here?

File: src/test/java/petStoreTests/user/CreateListOfUsersTest.java [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+        ResponseInfo response = UserClient.postUserList(users);
+
+        Assert.assertEquals(response.getCode(), HttpStatus.SC_OK);
+        Assert.assertEquals(response.getType(), DataSet.messageUnknownResponse().getType());
+        Assert.assertEquals(response.getMessage(), DataSet.messageUnknownResponse().setMessage("ok").getMessage());
+
+        if (users.size() != 0){
+            for (User user : users) {
+                User actualUser = UserClient.getUserByUsername(user.getUsername());
+                Assert.assertEquals(actualUser.getUsername(), user.getUsername());
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/CreateListOfUsersTest.java]: Same here - only verifying the username?

File: src/test/java/petStoreTests/user/FindUserNegativeTest.java [hunk 21]
```diff
@@ -0,0 +1,42 @@
...
+        Assert.assertEquals(response.getType(), DataSet.messageNotFoundResponse().getType());
+        Assert.assertEquals(response.getMessage(),
+                            DataSet.messageNotFoundResponse().setMessage("User not found").getMessage());
+        LOGGER.info("END TEST");
+    }
+
+    @DataProvider
+    public Object[][] testData() {
+        return new Object[][]{
+            {DataSet.addingUser().setUsername("nonExisting")},
```
Comment (andyyarych) [inline src/test/java/petStoreTests/user/FindUserNegativeTest.java]: Why do you use data provider here if there's only one case? Either add more input data or remove data provider

Comment (stadnytskyivp) [reply src/test/java/petStoreTests/user/FindUserNegativeTest.java]: ok I will delete it

File: src/main/java/data/ReusableMethods.java [hunk 22]
```diff
@@ -1,26 +1,35 @@
...
     public static String getBigData() throws IOException {
         Properties properties = new Properties();
         FileInputStream fis = new FileInputStream(System.getProperty("user.dir") +
             "/src/main/resources/log4j.properties");
         properties.load(fis);
         return properties.toString();
     }
+
+    public static void compareUsers(User newUser, User expectedUser){
+        Assert.assertNotNull(newUser);
```
Comment (andyyarych) [inline src/main/java/data/ReusableMethods.java]: Verify that id is not null as well

**Review by andyyarych — APPROVED** [review]

**Pull request merged**